firmware/led-spiout.c

```diff
@@ -21,55 +21,51 @@
  * shouldn't matter.
  */
 
+#define BRIGHTNESS_MASK 0b11100000
 
-uint8_t led_spi_frequency = LED_SPI_FREQUENCY_DEFAULT;
+// LED SPI start frame: 32 zero bits
+#define LED_START_FRAME_BYTES 4
+
+// LED SPI end frame: 32 zero bits + (NUM_LEDS / 2) bits
+// The pwm end frame needs to be 32 bits of 0 for SK9822 based LEDS
+// After that, we need at num_leds/2 more bits of 0
+// For up to 64 LEDs, that means 64 bits of 0
+// https://cpldcpu.wordpress.com/2016/12/13/sk9822-a-clone-of-the-apa102/
+// https://cpldcpu.wordpress.com/2014/11/30/understanding-the-apa102-superled/
+#define LED_END_FRAME_BYTES (4 + (NUM_LEDS / 2 / 8))
+
+#define ENABLE_LED_WRITES SPCR |= _BV(SPIE);
+#define DISABLE_LED_WRITES SPCR &= ~_BV(SPIE);
+
+// Using DISABLE_INTERRUPTS could cause interrupt recursion because led_update
+// functions are called from TWI callbacks which run within TWI interrupt. So
+// disable only LED SPI transfer interrupt.
+//
+// And to avoid redundancy, leave re-ENABLE_LED_WRITES to led_data_ready().
+#define PROTECT_LED_WRITES(...)  do { DISABLE_LED_WRITES; __VA_ARGS__ /*ENABLE_LED_WRITES;*/ } while (0)
+
```

Comment:
because all `PROTECT_LED_WRITES` calls are followed by `led_data_ready()` which do the `ENABLE_LED_WRITES`. `led_data_ready()` could be moved inside `PROTECT_LED_WRITES`...